src/which.js

```diff
@@ -45,74 +39,67 @@ function _which(options, cmd) {
+    noMatches: options.all ? [] : null,
+    value: [],
+  };
 
   // No relative/absolute paths provided?
-  if (cmd.search(/\//) === -1) {
+  if (cmd.indexOf('/') === -1) {
+    // Assume that there are no extensions to append to queries (this is the
+    // case for unix)
+    var pathExtArray = [''];
+    if (common.platform === 'win') {
+      // In case the PATHEXT variable is somehow not set (e.g.
+      // child_process.spawn with an empty environment), use the XP default.
+      var pathExtEnv = process.env.PATHEXT || XP_DEFAULT_PATHEXT;
+      pathExtArray = splitPath(pathExtEnv.toUpperCase());
+    }
+
     // Search for command in PATH
-    pathArray.forEach(function (dir) {
-      if (!all && where) return; // already found it
+    for (var k = 0; k < pathArray.length; k++) {
+      // already found it
+      if (queryMatches.hasFoundSomething() && !options.all) break;
 
-      var attempt = path.resolve(dir, cmd);
+      var attempt = path.resolve(pathArray[k], cmd);
 
       if (common.platform === 'win') {
         attempt = attempt.toUpperCase();
+      }
 
-        // In case the PATHEXT variable is somehow not set (e.g.
-        // child_process.spawn with an empty environment), use the XP default.
-        var pathExtEnv = process.env.PATHEXT || XP_DEFAULT_PATHEXT;
-        var pathExtArray = splitPath(pathExtEnv.toUpperCase());
-        var i;
-
-        // If the extension is already in PATHEXT, just return that.
-        for (i = 0; i < pathExtArray.length; i++) {
-          var ext = pathExtArray[i];
-          if (attempt.slice(-ext.length) === ext && checkPath(attempt)) {
-            if (all) {
-              where.push(attempt);
-            } else {
-              where = attempt;
-              return;
-            }
-          }
+      // Cycle through the PATHEXT variable
+      for (var i = 0; i < pathExtArray.length; i++) {
+        var ext = pathExtArray[i];
+        // If the argument passed in already has an extension in PATHEXT, just
+        // return that
+        // TODO(nate): replace with .endsWith(ext) once we use v4+
+        if (ext && attempt.slice(-ext.length) === ext && checkPath(attempt)) {
+          queryMatches.add(attempt);
+          break;
         }
 
-        // Cycle through the PATHEXT variable
-        var baseAttempt = attempt;
-        for (i = 0; i < pathExtArray.length; i++) {
-          attempt = baseAttempt + pathExtArray[i];
-          if (checkPath(attempt)) {
-            if (all) {
-              where.push(attempt);
-            } else {
-              where = attempt;
-              return;
-            }
-          }
-        }
-      } else {
-        // Assume it's Unix-like
-        if (checkPath(attempt)) {
-          if (all) {
-            where.push(attempt);
-          } else {
-            where = attempt;
-            return;
-          }
+        // Otherwise, see if it exists without the extension
+        var newAttempt = attempt + ext;
```

Comment:
This is a bit confusing. We're checking if it exists "without the extension", but then we add the extension?